### ISSUE TITLE
HypreParMatrix data depends on the ParBilinearForm that assembled it.

### ISSUE THREAD

**Issue by dylan-copeland** [mfem/mfem#793]
When a ParBilinearForm is used to create a HypreParMatrix, the resulting HypreParMatrix has data that depends on the ParBilinearForm instance, which is lost if the ParBilinearForm is destroyed. This can be reproduced with the mfem/examples, e.g. ex1p. As a result, the user must be careful not to delete the ParBilinearForm or let it go out of scope, while still using the HypreParMatrix, but that is not obvious. Either this should be fixed so that the HypreParMatrix owns all the data independently of ParBilinearForm, or a warning should be added in comments to ParBilinearForm. 

Here are some details on how this happens.

1) ParBilinearForm::Assemble() gets called, which calls BilinearForm::Assemble(), which allocates BilinearForm::mat (of type SparseMatrix*).

2) ParBilinearForm::FormSystemMatrix gets called with a HypreParMatrix as the argument to be set. Since mat is defined, it calls ParallelAssemble(p_mat, mat);

3) ParBilinearForm::ParallelAssemble calls OperatorHandle::MakeSquareBlockDiag, which calls HypreParMatrix(comm, glob_size, row_starts, diag); with diag = mat.

4) The constructor for HypreParMatrix uses HypreParMatrix::CopyCSR to set the I, J, and data pointers for p_mat to the data pointers for mat.

5) p_mat is returned as the resulting HypreParMatrix.

In summary, it appears that the serial BilinearForm SparseMatrix is used to store local data for the parallel HypreParMatrix returned by ParBilinearForm. Thus the HypreParMatrix is not destroyed when 
ParBilinearForm is destroyed, but its data is.

**Comment by v-dobrev** [comment]
If I understand the issue correctly, you want to destroy the `ParBilinearForm` but keep the `HypreParMatrix` that it created. Is that correct? If that is the case, we need to add a method that lets you take ownership of the `HypreParMatrix` from the `ParBilinearForm`. That method simply needs to set the ownership flag of `p_mat` to false (with `SetOperatorOwner()`) and return either directly the pointer to the `HypreParMatrix` or copy the pointer to an `OperatorHandle` (setting the ownership flag of that to true).

**Comment by dylan-copeland** [comment]
@v-dobrev Yes, I mean that HypreParMatrix should not become invalid when ParBilinearForm is destroyed. Unless it is clearly documented, I would not expect the user to know that the HypreParMatrix depends on the ParBilinearForm, so the user should not need to call a method to take ownership. It would be simpler for the user if ParBilinearForm never owns data in the HypreParMatrix, but your suggestion would be fine with documentation.

**Comment by NoseKnowsAll** [comment]
I ran into this same issue as well. Basically, I agree with Dylan that it doesn't make sense for the `HypreParMatrix` to actually depend on the underlying `ParBilinearForm`. The function call `FormSystemMatrix()` appears to create a valid matrix that you can take and use wherever, and I think that should be the intended behavior. 

As an example, I think the following code should provide a valid `A` with or without the brackets. This would allow for a user to write a function to setup the spaces, integrators, etc. and just return the final matrix.

```
   HypreParMatrix A;
   {
      ParBilinearForm a(&fes);
      a.AddDomainIntegrator(new DiffusionIntegrator(one));
      a.Assemble();
      a.FormSystemMatrix(ess_tdof_list, A);
   }
```

As-is, it appears that the matrix is independent of the bilinearform, but actually the internal `*(A.A->diag)` matrix becomes garbage when the `ParBilinearForm` goes out of scope. There might be more internals that are dependent on the form, but this is the one I've come across as a user from the outside looking in.

**Comment by ben-s-southworth** [comment]
I agree, I find this very confusing. It actually says in the documentation "The returned matrix has to be deleted by the caller," which seems to suggest that the matrix does not depend on the bilinear form. It also seems that the bilinear form can depend on other things, so even if the bilinear form is a pointer and still in scope, something else has gone out of scope, which effects the bilinear form, and then messes up the HypreParMatrix (still trying to track down what's going on in our code).

**Comment by tzanio** [comment]
@v-dobrev what do you think?

**Comment by v-dobrev** [comment]
The assembled `HypreParMatrix` with the imposed b.c. is currently owned by the `ParBilinearForm`. The `HypreParMatrix` object given to the methods `FormLinearSystem`/`FormSystemMatrix` just stores a pointer (which it does not own) to the same `hypre_ParCSRMatrix` (this is the actual hypre type) as the `HypreParMatrix` owned by the `ParBilinearForm`. It is easy to just transfer the ownership flag inside the method `FormSystemMatrix`. However, this may have implications for existing codes, e.g. it may break things. A better option, I think, is to modify the `HypreParMatrix` object so that it shares the pointer to `hypre_ParCSRMatrix` with other `HypreParMatrix` objects. This way the order of destruction should not matter -- either `HypreParMatrix` object will remain valid even if the other one is destroyed.

**Comment by dylan-copeland** [comment]
It appears that this is also an issue for BilinearForm and SparseMatrix, as BilinearForm::FormSystemMatrix makes the argument A a reference to mat owned by BilinearForm, which is deleted by the BilinearForm destructor. This also affects BilinearForm::FormLinearSystem.

**Comment by NoseKnowsAll** [comment]
For what it's worth, this is somewhat of an issue for matrix-free operators as well. In fact, for matrix-free operators, it makes more sense to always keep the underlying BilinearForm and not use the "matrix" in isolation.

Therefore, to be consistent with all use-cases, I now believe that we probably need to modify the language in the description, more than the functionality itself. As long as it's obvious that these matrices will be invalidated if/when the BilinearForm goes out of scope, there wouldn't be much of an issue. What do y'all think?

**Comment by ben-s-southworth** [comment]
I think there should be a straightforward way to keep the parallel matrix constructed by the bilinear form, even if the bilinear form is deleted or goes out of scope. Because there are plenty of situations where you need the matrix but not the bilinear form. I don't totally understand how Veselin's idea will work, but as long as the HypreParMatrix is/can be still valid after the bilinearform destructs, that is okay.

**Comment by theownerof** [comment]
Yo what the fuck is you guys talking about why the fuck is this on my phone anyway

**Comment by dylan-copeland** [comment]
Branch `hypredata` contains my first attempt at implementing a fix similar to what @v-dobrev suggested. The idea is to use `shared_ptr` for the `I, J, A` data in `SparseMatrix`. This data is part of `HypreParMatrix` (after `CopyCSR` is called), but it gets deleted when `ParBilinearForm` is destroyed. By maintaining a `shared_ptr` for each of these 3 arrays in `SparseMatrix` and `HypreParMatrix`, we should be able to prevent the deletion of the data until the final object using them is destroyed. 

My goal was to see if I could get something working in a few hours, and I did not reach that goal. I added a unit test for this new capability, and a couple of lines in `ex1p` to delete the `ParBilinearForm` and see if the linear solve still works. The tests fail.

I am not sure if it is worth much more (of my) time to get this working. Maybe I will revisit this some time. In any case, I would appreciate any help or feedback. Some questions to answer are: 

1. Is this the best approach?
2. If so, how can this branch be fixed?
3. Is it acceptable to make so many changes to `linalg` for this purpose, and will those changes break anything?

**Comment by v-dobrev** [comment]
Reading my post again (https://github.com/mfem/mfem/issues/793#issuecomment-549630950), I think the second option was to use
```c++
   shared_ptr<hypre_ParCSRMatrix> A;
```
instead of
```c++
   hypre_ParCSRMatrix *A;
```
inside the class `HypreParMatrix`. This will probably work fine in parallel. However, in the serial case (if we want similar behavior) this becomes a bit harder as Dylan found out -- it will require sharing the parts of the `SparseMatrix` (like the `I`, `J`, and `data` arrays) with another `SparseMatrix`.

Dylan, I think in the parallel case you are thinking of the temporary block-diagonal hypre matrix before the RAP with the parallel prolongation -- this is the hypre matrix that reuses pointers from the `SparseMatrix` representing the locally assembled block. Since this is a temporary hypre matrix, there is no need to worry about sharing the pointer ownerships there. The result of the RAP is a different matrix that is owned by the `ParBilinearForm`. The first step is performed here (the comment is out of date):
https://github.com/mfem/mfem/blob/99cf263ee490cd0431d7b2b798e3070c65a44d81/fem/pbilinearform.cpp#L135-L137

and the second one, here:
https://github.com/mfem/mfem/blob/99cf263ee490cd0431d7b2b798e3070c65a44d81/fem/pbilinearform.cpp#L179

A simpler solution is probably what I suggested in a previous comment (https://github.com/mfem/mfem/issues/793#issuecomment-495680186) but it seems like @ben-s-southworth and @NoseKnowsAll didn't like that approach.

**Comment by v-dobrev** [comment]
> I think there should be a straightforward way to keep the parallel matrix constructed by the bilinear form, even if the bilinear form is deleted or goes out of scope. Because there are plenty of situations where you need the matrix but not the bilinear form. I don't totally understand how Veselin's idea will work, but as long as the HypreParMatrix is/can be still valid after the bilinearform destructs, that is okay.
> 
> @v-dobrev Is there a current work around for this? E.g., I construct a mass matrix
> 
> ```
>    /// Mcv
>    ParBilinearForm mcv_bf(fes_);
>    mcv_bf.AddDomainIntegrator(new MassIntegrator(*rhoCv_pcf_));
>    mcv_bf.Assemble();
>    mcv_bf.Finalize();
>    HypreParMatrix *Mcv = mcv_bf.ParallelAssemble(); 
> ```
> 
> in a function where I need to save/return the HypreParMatrix Mcv for nonlinear iterations. I don't need/want the extra ParBilinear form hanging around, but I believe if `mcv_bf` above is created (and destructred) in this auxiliary function, the HypreParMatrix will be too?

@ben-s-southworth, you probably already figured this out but let me try to answer you question for completeness. When you call `mcv_bf.ParallelAssemble()` the returned matrix is not owned by the `ParBilinearForm`, so destroying it will not destroy the `HypreParMatrix`. I think the only possible dependence of the returned `HypreParMatrix` are the parallel offsets for the rows/columns -- I think these are constructed and owned by the `ParFiniteElementSpace`, so if that is destroyed, the matrix will become invalid. You can call the method `HypreParMatrix::CopyRowStarts` (and `HypreParMatrix::CopyColStarts` for `ParMixedBilinearForm`s) to make sure these offsets are copied and owned by the `HypreParMatrix`.
